python/cudf/dataframe/multiindex.py

```diff
@@ -85,8 +94,14 @@ def _validate_levels_and_codes(self, levels, codes):
                                  'than maximum level size at this position')
 
     def copy(self, deep=True):
-        mi = MultiIndex(self.levels.copy(),
-                        self.codes.copy(deep))
+        if hasattr(self, '_source_data'):
+            mi = MultiIndex(source_data=self._source_data)
+            if self._levels is not None:
+                mi._levels = self._levels.copy()
+            if self._codes is not None:
+                mi._codes = self._codes.copy(deep)
+        else:
+            mi = MultiIndex(self.levels, self.codes)
```

Comment:
If `_source_data` is None and deep=True, won't this construct a shallow copy because it will use those by reference?

Reply:
why are you always right